src/HashMsg.chpl

```diff
@@ -0,0 +1,97 @@
+module HashMsg {
+  use Reflection;
+  use ServerErrors;
+  use ServerErrorStrings;
+  use Logging;
+  use Message;
+
+  use MultiTypeSymbolTable;
+  use MultiTypeSymEntry;
+  use CommAggregation;
+  use ServerConfig;
+  use SegmentedString;
+  use AryUtil;
+  use UniqueMsg;
+
+  private config const logLevel = ServerConfig.logLevel;
+  private config const logChannel = ServerConfig.logChannel;
+  const hmLogger = new Logger(logLevel, logChannel);
+
+  proc categoricalHash(categoriesName: string, codesName: string, st: borrowed SymTab) throws {
+    var categories = getSegString(categoriesName, st);
+    var codes = toSymEntry(getGenericTypedArrayEntry(codesName, st), int);
+    // hash categories first
+    var hashes = categories.siphash();
+    // then do expansion indexing at codes
+    ref ca = codes.a;
+    var expandedHashes: [ca.domain] (uint, uint);
+    forall (eh, c) in zip(expandedHashes, ca) with (var agg = newSrcAggregator((uint, uint))) {
+      agg.copy(eh, hashes[c]);
+    }
+    var hash1 = makeDistArray(ca.size, uint);
+    var hash2 = makeDistArray(ca.size, uint);
+    forall (h, h1, h2) in zip(expandedHashes, hash1, hash2) {
+      (h1,h2) = h:(uint,uint);
+    }
+    return (hash1, hash2);
+  }
+
+  proc categoricalHashMsg(cmd: string, msgArgs: borrowed MessageArgs, st: borrowed SymTab): MsgTuple throws {
+    var pn = Reflection.getRoutineName();
+    var repMsg: string;
+    const objtype = msgArgs.getValueOf("objType").toUpper(): ObjType;
+    select objtype {
+      when ObjType.CATEGORICAL {
+        // check to make sure symbols defined
+        const categoriesName = msgArgs.getValueOf("categories");
+        const codesName = msgArgs.getValueOf("codes");
+        st.checkTable(categoriesName);
+        st.checkTable(codesName);
+        var (upper, lower) = categoricalHash(categoriesName, codesName, st);
+        var upperName = st.nextName();
+        st.addEntry(upperName, new shared SymEntry(upper));
+        var lowerName = st.nextName();
+        st.addEntry(lowerName, new shared SymEntry(lower));
+        repMsg = "created " + st.attrib(upperName) + "+created " + st.attrib(lowerName);
```

Comment:
Just noting that if we do the update to JSON return structure this will need to update.